docs/design/2022-06-22-foreign-key.md

```diff
@@ -0,0 +1,548 @@
+
+```sql
+create table t1 (id int key,a int, index(a));
+create table t2 (id int key,a int, foreign key fk(a) references t1(id) ON DELETE CASCADE);
+rename table t1 to t11;
+alter table t11 change column id id1 int;
+show create table t2\G
+***************************[ 1. row ]***************************
+    Table        | t2
+Create Table | CREATE TABLE `t2` (
+                                     `id` int NOT NULL,
+                                     `a` int DEFAULT NULL,
+                                     PRIMARY KEY (`id`),
+    KEY `fk` (`a`),
+    CONSTRAINT `t2_ibfk_1` FOREIGN KEY (`a`) REFERENCES `t11` (`id1`) ON DELETE CASCADE
+    ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci
+```
+
+### Modify Column
+
+Modify column which used by foreign key will be rejected.
+
+```sql
+> alter table t1 change column id id1 bigint;
+(3780, "Referencing column 'a' and referenced column 'id1' in foreign key constraint 't2_ibfk_1' are incompatible.")
+```
+
+MySQL modify column problem: https://www.percona.com/blog/2019/06/04/ddl-queries-foreign-key-columns-MySQL-pxc/
+
+What if the user really need to modify column type, such as from `INT` to `BIGINT`. Maybe we can offer an variable such as `alter-foreign-keys-method=auto`,
+then when user modify the column type, TiDB will auto modify the related foreign key column's type. For easy implementation and reduce risk, maybe only support modify column type which doesn't need to reorg table row data.
+
+## DML Technical Design
+
+### DML On Child Table
+
+On Child Table Insert Or Update, need to Find FK column value whether exist in reference table:
+
+1. Get reference table info by table name.
+2. Get related fk index of reference table.
+3. tiny optimize, check fk column value exist in reference table cache(map[string(index_key)]struct).
+3. Get related row in reference.
+  - Construct index key and then use snapshot `Iter` and `Seek` API to scan. If the index is unique and only contain
+    foreign key columns, use snapshot `Get` API.
+    - `Iter` default scan batch size is 256, need to set 2 to avoid read unnecessary data.
+4. compact column value to make sure exist.
+5. put column value into reference fk column value cache.
+
+check order should check unique/primary key constrain first:
+
+```sql
+test> create table t1 (id int key,a int, index(a));
+test> create table t2 (id int key,a int, foreign key fk(a) references t1(id) ON DELETE CASCADE);
+test> insert into t1 values (1, 1);
+test> insert into t2 values (1, 1);
+test> insert into t2 values (1, 2);
+(1062, "Duplicate entry '1' for key 't2.PRIMARY'")
+test> insert ignore into t2 values (1, 2);
+Query OK, 0 rows affected
+```
+
+Load data should also do foreign key check, but report warning instead error:
+
+```sql
+create table t1 (id int key,a int, index(a));
+create table t2 (id int key,a int, foreign key fk(a) references t1(id) ON DELETE CASCADE);
```

Comment:
better add a placement rule here, let t1 & t2 which has the foreign-key bound ship table's leader scheduled in a single kv node.

Reply:
I also think about the mechanism of a kind of inherited table means from postgre, letting table t2's encoded in the range of table t1's scope.